### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Haremun**
Logger,
Board indexing,
Horizontal win checker

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -20,6 +20,9 @@
         <maven.resources.encoding>${project.build.sourceEncoding}</maven.resources.encoding>
         <version.plugin.maven.jacoco>0.8.5</version.plugin.maven.jacoco>
         <version.maven>3.6.3</version.maven>
+        <version.plugin.maven.enforcer>3.0.0-M2</version.plugin.maven.enforcer>
+
+        <argumentA>15</argumentA>
     </properties>
```
Comment (patrycja93) [inline pom.xml:26]: It is not correct name ;)

File: pom.xml [hunk 2]
```diff
@@ -29,11 +32,87 @@
...
+                                </rule>
+                            </rules>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+
         <pluginManagement>
             <plugins>
```
Comment (patrycja93) [inline pom.xml]: It is useful if you have a few modules in project.

File: src/main/java/com/bieganski/ox/App.java [hunk 3]
```diff
@@ -1,22 +1,28 @@
...
-      
-        PositionCalculator positionCalculator = new PositionCalculator(new PositionValidator(),10);
-      
-        String input = userInterface.askForInput();
-        if (new PositionValidator().checkString(input)) {
-            board.addField(new Field(positionCalculator.calculatePosition(input), Symbol.X));
-        }
+        //TODO: change const values ,create game setup
+        GameFlow gameFlow = new GameFlow(userInterface, Integer.parseInt(args[0]));
+        new Thread(gameFlow).start();
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/App.java:29]: Integer.parseInt(args[0]) can return exception. Maybe you should handle it before put to constructor ?

File: src/main/java/com/bieganski/ox/GameFlow.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    GameFlow(UserInterface userInterface, int boardSize) {
+        this.boardSize = boardSize;
+        this.userInterface = userInterface;
+
+        players.add(Symbol.X);
+        players.add(Symbol.O);
+
+        initializeBoard();
+    }
+
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/GameFlow.java:25]: You have logic in constructor. You should think about factory method.

File: src/main/java/com/bieganski/ox/GameFlow.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    public void run() {
+        PositionCalculator positionCalculator = new PositionCalculator(new PositionValidator(), boardSize);
+        Symbol currentSymbol = players.get(0);
+        while (!win) {
+            userInterface.println("Input row and column:");
+            String input = userInterface.askForInput();
+            boolean added = board.addField(new Field(positionCalculator.calculatePosition(input), currentSymbol));
+            if (added)
+                currentSymbol = changePlayer(currentSymbol);
+        }
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/GameFlow.java:44]: Even if you have one line in if statement you should use {} .

File: src/main/java/com/bieganski/ox/Player.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+public class Player {
+    private String name;
+    private int score;
+    private Symbol symbol;
+
+    public Player(String name) {
+        this.name = name;
+    }
+}
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/Player.java:11]: You have variables score and symbol which are not using :(

File: src/main/java/com/bieganski/ox/PositionCalculator.java [hunk 7]
```diff
@@ -15,6 +15,11 @@ int calculatePosition(String userInput) {
         String[] coordinates = userInput.split(" ");
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/PositionCalculator.java:15]: Maybe the better idea would be create other class to check user input.

File: src/main/java/com/bieganski/ox/PositionCalculator.java [hunk 8]
```diff
@@ -15,6 +15,11 @@ int calculatePosition(String userInput) {
...
         String column = coordinates[1];
-        return Integer.parseInt(row) * boardSize + Integer.parseInt(column);
+        try {
+            return Integer.parseInt(row) * boardSize + Integer.parseInt(column);
+        } catch (NumberFormatException ex){
+            App.LOG.warn("User provided incorrect number");
+            return -1;
+        }
     }
 }
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/PositionCalculator.java:25]: The same as up. Think about new class to check user input instead make it inside the method.

File: src/main/java/com/bieganski/ox/painters/ConsoleBoardPainter.java [hunk 9]
```diff
@@ -19,8 +30,11 @@ public String paintBoard(TreeSet<Field> fieldsWithValues, int boardSide) {
...
 
-            if ((position + 1) % boardSide == 0)
+            if ((position + 1) % boardSide == 0 && position + 1 != boardSide * boardSide) {
                 stringBuilder.append("\n");
+                counter++;
+                indexCreator.nextIndex(stringBuilder, counter);
+            }
         }
 
         return stringBuilder.toString();
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/painters/ConsoleBoardPainter.java:40]: This method is loo long.

File: src/main/java/com/bieganski/ox/painters/IndexCreator.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    void nextIndex(StringBuilder stringBuilder, int index) {
+
+        if (index < 10)
+            stringBuilder.append(String.format(" %s ", index));
+        else if (index < 100)
+            stringBuilder.append(String.format("%s ", index));
+        else
+            stringBuilder.append(String.format("%s", index));
+    }
+
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/painters/IndexCreator.java:27]: Again if statement without {}. Be carefully when use it.

File: src/main/java/com/bieganski/ox/win_checkers/ConsoleJudge.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+public class ConsoleJudge implements Judge {
+    private List<WinChecker> checkers = new ArrayList<>();
+    private int boardSize;
+
+    public ConsoleJudge(int boardSize) {
+        checkers.add(new HorizontalWinChecker());
+        this.boardSize = boardSize;
+    }
+
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/win_checkers/ConsoleJudge.java]: Again you have logic in constructor.

File: src/main/java/com/bieganski/ox/win_checkers/HorizontalWinChecker.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+            field = correct.first();
+            nextField = fieldsWithValue.lower(field);
+            counter = counter + appendFieldIfCorrect(field, nextField, correct);
+
+
+        }
+        return correct.size() >= winSize;
+    }
+
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/win_checkers/HorizontalWinChecker.java:30]: This method could be shorter.

File: src/main/java/com/bieganski/ox/win_checkers/Judge.java [hunk 13]
```diff
@@ -0,0 +1,6 @@
+package com.bieganski.ox.win_checkers;
+
+import com.bieganski.ox.BoardListener;
+
+public interface Judge extends BoardListener {
+}
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/win_checkers/Judge.java:32]: You have empty interface.

File: src/test/java/com/bieganski/ox/HorizontalWinCheckerTest.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        //Given
+        TreeSet<Field> fields = new TreeSet<>();
+        fields.add(new Field(0, Symbol.X));
+        fields.add(new Field(1, Symbol.O));
+        HorizontalWinChecker checker = new HorizontalWinChecker();
+        //When
+        boolean isWin = checker.checkWin(fields, new Field(3, Symbol.X), 10);
+        //Then
+        assertFalse(isWin);
+    }
```
Comment (patrycja93) [inline src/test/java/com/bieganski/ox/HorizontalWinCheckerTest.java:65]: Think about data provider in tests.

**Review by patrycja93 — APPROVED** [review]
You have a little errors which are not big but think about my signs.

File: src/main/java/com/bieganski/ox/Field.java [hunk 15]
```diff
@@ -12,14 +12,21 @@
...
+    public boolean isOnPosition(int position) {
         return this.position == position;
     }
 
     boolean isInBounds(int bounds) {
         return 0 <= position && position < bounds;
     }
-  
+
+    public boolean areSymbolsEquals(Field field) {
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/Field.java]: areSymbolEqual - without "s" at the end

File: src/main/java/com/bieganski/ox/Field.java [hunk 16]
```diff
@@ -12,14 +12,21 @@
...
+        return this.symbol.equals(field.symbol);
+    }
+
+    public int distanceToField(Field field) {
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/Field.java]: distanceFrom() maybe?

File: src/main/java/com/bieganski/ox/Player.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package com.bieganski.ox;
+
+public class Player {
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/Player.java:3]: I love this class xD

File: src/main/java/com/bieganski/ox/painters/ConsoleBoardPainter.java [hunk 18]
```diff
@@ -1,15 +1,26 @@
...
+import com.bieganski.ox.Field;
 
 import java.util.TreeSet;
 
-class ConsoleBoardPainter implements BoardPainter {
+public class ConsoleBoardPainter implements BoardPainter {
+
+    //TODO: Too long function
     @Override
     public String paintBoard(TreeSet<Field> fieldsWithValues, int boardSide) {
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/painters/ConsoleBoardPainter.java:11]: To long!

File: src/main/java/com/bieganski/ox/win_checkers/ConsoleJudge.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    @Override
+    public void onBoardUpdate(TreeSet<Field> fieldsWithValue, Field addedField, int size) {
+        boolean win = false;
+        for (WinChecker checker :
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/win_checkers/ConsoleJudge.java]: Could be one line

File: src/main/java/com/bieganski/ox/win_checkers/HorizontalWinChecker.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        int counter = 0;
+
+        while (correct.size() < winSize && counter < 2) {
+            field = correct.last();
+            Field nextField = fieldsWithValue.higher(field);
+            counter = counter + appendFieldIfCorrect(field, nextField, correct);
+
+            field = correct.first();
+            nextField = fieldsWithValue.lower(field);
+            counter = counter + appendFieldIfCorrect(field, nextField, correct);
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/win_checkers/HorizontalWinChecker.java:24]: Duplicated code. Try to fix it

File: src/main/java/com/bieganski/ox/win_checkers/HorizontalWinChecker.java [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+            return 0;
+        }
+        return 1;
+    }
+
+    private boolean checkFieldsAreNotNull(Field current, Field next) {
+        return current != null && next != null;
+    }
+
+    private boolean isDistanceCorrectAndSymbolsAreEquals(Field current, Field next) {
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/win_checkers/HorizontalWinChecker.java]: I like this name :)

File: src/main/java/com/bieganski/ox/win_checkers/WinChecker.java [hunk 22]
```diff
@@ -0,0 +1,14 @@
+package com.bieganski.ox.win_checkers;
+
+import com.bieganski.ox.BoardListener;
+import com.bieganski.ox.Field;
+
+import java.util.TreeSet;
+
+/**
+ * Interface for judge classes
+ * @author Kamil Biegański
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/win_checkers/WinChecker.java]: I did not see this ;)

**Review by hongshui — APPROVED** [review]
I see a big improvement compared to previous versions :)

File: src/main/java/com/bieganski/ox/win_checkers/ConsoleJudge.java (hunk 19 above)
Comment (Haremun) [reply src/main/java/com/bieganski/ox/win_checkers/ConsoleJudge.java]: Great idea. Did it with stream and anyMatch

File: pom.xml (hunk 2 above)
Comment (Haremun) [reply pom.xml]: Removed

File: src/main/java/com/bieganski/ox/App.java (hunk 3 above)
Comment (Haremun) [reply src/main/java/com/bieganski/ox/App.java:29]: It's temporary

**Pull request merged**